lib/opencensus/trace/span.rb

```diff
@@ -0,0 +1,198 @@
+# Copyright 2017 OpenCensus Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module OpenCensus
+  module Trace
+    ##
+    # Span represents a span in a trace record. Spans are contained in
+    # a trace and arranged in a forest. That is, each span may be a root span
+    # or have a parent span, and may have zero or more children.
+    #
+    class Span
+      ##
+      # A unique identifier for a trace. All spans from the same trace share
+      # the same `trace_id`. The ID is a 16-byte value represented as a
+      # hexadecimal string.
+      #
+      # @return [String]
+      #
+      attr_reader :trace_id
+
+      ##
+      # A unique identifier for a span within a trace, assigned when the span
+      # is created. The ID is an 8-byte value represented as a hexadecimal
+      # string.
+      #
+      # @return [String]
+      #
+      attr_reader :span_id
+
+      ##
+      # The `span_id` of this span's parent span. If this is a root span, then
+      # this field must be empty. The ID is an 8-byte value represented as a
+      # hexadecimal string.
+      #
+      # @return [String]
+      #
+      attr_reader :parent_span_id
+
+      ##
+      # The name of this span.
+      #
+      # @return [OpenCensus::Trace::TruncatableString]
+      #
+      attr_reader :name
+
+      ##
+      # The starting timestamp of this span in UTC, or `nil` if the
+      # starting timestamp has not yet been populated.
+      #
+      # @return [Time, nil]
+      #
+      attr_reader :start_time
+
+      ##
+      # The ending timestamp of this span in UTC, or `nil` if the
+      # ending timestamp has not yet been populated.
+      #
+      # @return [Time, nil]
+      #
+      attr_reader :end_time
+
+      ##
+      # The properties of this span.
+      #
+      # @return [Hash{String => String}]
+      #
+      attr_reader :attributes
+
+      ##
+      # The number of attributes that were discarded. Attributes can be
+      # discarded because their keys are too long or because there are too
+      # many attributes. If this value is 0, then no attributes were dropped.
+      attr_reader :dropped_attributes_count
+
+      ##
+      # A stack trace captured at the start of the span.
+      #
+      # @return [Thread::Backtrace::Location[]]
```

Comment:
I believe the preferred type annotation syntax is `Array{Thread::Backtrace::Location}`. Same for time_events and links below.

Reply:
Ok, looks like the YARD format is Array<subtype>, so Array<Thread::Backtrace::Location> and Hash<keytype, (childtype1, childtype2)> for applicable types. I'll fix these.

Reply:
Oh dear... yard specifies angle brackets? I think we've been using curly braces in google-cloud-ruby (maybe to prevent markdown from thinking it's an html tag).

Reply:
Okay, according to https://yardoc.org/types.html Array specifically uses angle brackets, while Hash specifically uses curly braces. :-/